### PR TITLE
Record langversion diagnostics

### PR DESCRIPTION
**Pull request by RikkiGibson**
Related to #45900

Fixes https://github.com/dotnet/roslyn/projects/57#card-39518530 (this was already marked Done, but seems like it was only working when we incidentally were getting a diagnostic for top-level-statements which resemble record declarations.)

The idea here is that if 'record' fails to bind to a type in C# 8 or less, in a position where we would have parsed the declaration as a record in C# 9, such as a field, property, or ordinary method, local variable, or local function (for top level statement scenarios) we will give a LangVersion error in order to help you upgrade your project in the IDE.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
Is this PR improving https://github.com/dotnet/roslyn/issues/45900?

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs [hunk 1]
```diff
@@ -1964,5 +1965,13 @@ internal static int TypeToIndex(this TypeSymbol type)
                 default: return -1;
             }
         }
+
+        internal static void CheckRecordLangVersion(this TypeSymbol type, DiagnosticBag diagnostics, TypeSyntax syntax, Location location)
+        {
+            if (type.IsErrorType() && syntax is IdentifierNameSyntax name && name.Identifier.Text == SyntaxFacts.GetText(SyntaxKind.RecordKeyword))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs:1971]: >type.IsErrorType() [](start = 16, length = 18)

Would it be possible to be more precise in detecting the situation when the type is not found?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs:1971]: >Text [](start = 93, length = 4)

Why use Text? I think we usually use Value and ValueText properties, or something like that.

**Comment by RikkiGibson** [comment]
> Is this PR improving #45900?

From the linked issue:

> From our discussion, it would be good to at least provide a better/additional diagnostic, but ideally we would reparse as a record type.

With this change, we provide a LangVersion diagnostic for records in more record scenarios, while continuing to provide the LangVersion diagnostic for top level statements in some (fewer) record scenarios.

There may also be a reasonable method of avoiding giving the LangVersion diagnostic for top-level-statements when the statement would parse as a record in C# 9.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 2]
```diff
@@ -830,6 +830,8 @@ private TypeWithAnnotations BindVariableTypeWithAnnotations(CSharpSyntaxNode dec
                     // Keep processing it as a non-const local.
                     isConst = false;
                 }
+
+                declType.Type.CheckRecordLangVersion(diagnostics, typeSyntax, typeSyntax.Location);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:834]: >declType.Type.CheckRecordLangVersion(diagnostics, typeSyntax, typeSyntax.Location); [](start = 16, length = 83)

Where are we making sure that we are not inside a method body? It looks like ```record x;``` inside a body is never parsed as a record declaration.

File: src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs [hunk 3]
```diff
@@ -267,6 +267,8 @@ internal void ComputeReturnType()
                 diagnostics.Add(ErrorCode.ERR_MethodReturnCantBeRefAny, returnTypeSyntax.Location, returnType.Type);
             }
 
+            returnType.Type.CheckRecordLangVersion(diagnostics, returnTypeSyntax, returnTypeSyntax.Location);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs:270]: >returnType.Type.CheckRecordLangVersion(diagnostics, returnTypeSyntax, returnTypeSyntax.Location); [](start = 12, length = 97)

Where are we making sure we are not inside a method/local function/lambda body?

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs:1971]: Perhaps `ErrorTypeSymbol.ResultKind`?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs [hunk 4]
```diff
@@ -82,6 +82,8 @@ protected void TypeChecks(TypeSymbol type, DiagnosticBag diagnostics)
                 diagnostics.Add(ErrorCode.ERR_VolatileStruct, this.ErrorLocation, this, type);
             }
 
+            type.CheckRecordLangVersion(diagnostics, TypeSyntax, TypeSyntax?.Location);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs:85]: >TypeSyntax?.Location [](start = 65, length = 20)

Why conditional access? It doesn't look like the helper is expecting null as location.

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs [hunk 5]
```diff
@@ -1964,5 +1965,13 @@ internal static int TypeToIndex(this TypeSymbol type)
                 default: return -1;
             }
         }
+
+        internal static void CheckRecordLangVersion(this TypeSymbol type, DiagnosticBag diagnostics, TypeSyntax syntax, Location location)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs:1969]: >Location location [](start = 120, length = 17)

It looks like all the consumers are passing location of the previous parameter. Consider removing this parameter and requesting location inside this helper when we actually need it.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs:1969]: >CheckRecordLangVersion [](start = 29, length = 22)

Are all call sites for this helper covered by tests?

**Comment by AlekseyTs** [comment]
It looks like we are missing tests for scenarios inside a method/local function/lambda body. And tests with escaped identifiers, which should never trigger the upgrade offer.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 3)

**Comment by RikkiGibson** [comment]
I think the window for 16.8 is too small at this point. Let's get this into 16.9 instead.

**Comment by RikkiGibson** [comment]
Unfortunately I think records have been out long enough that the effort of finishing this PR is not justified.

**Pull request closed without merging**